src/v/serde/parquet/encoding.cc

```diff
@@ -18,81 +18,6 @@
 #include <climits>
 
 namespace serde::parquet {
-
-iobuf encode_plain(const chunked_vector<boolean_value>& vals) {
-    iobuf buf;
-    size_t i = 0;
-    while (i < vals.size()) {
-        size_t pass_size = std::min<size_t>(vals.size() - i, CHAR_BIT);
-        uint8_t tmp = 0;
-        size_t shift = 0;
-        for (; shift < pass_size; ++i, ++shift) {
-            if (vals[i].val) {
-                tmp |= 1U << shift;
-            }
-        }
-        buf.append(&tmp, 1);
-    }
-    return buf;
-}
-
-iobuf encode_plain(const chunked_vector<int32_value>& vals) {
-    iobuf buf;
-    for (const auto& v : vals) {
-        int32_t i = ss::cpu_to_le(v.val);
-        // NOLINTNEXTLINE(*reinterpret-cast*)
-        buf.append(reinterpret_cast<const uint8_t*>(&i), sizeof(int32_t));
-    }
-    return buf;
-}
-
-iobuf encode_plain(const chunked_vector<int64_value>& vals) {
```

Comment:
The original reason I had done this was so that it could be vectorized. Would that be faster/better than what we are doing?

I am not sure if the current implementation is vectorized, but I am assuming not because of our custom containers and the iobuf work being done.

Would an approach of working in chunks that can be vectorized be faster?

Reply:
By vectorized do you mean the `cpu_to_le` and/or the copying? For `cpu_to_le` it should be close to a no-op on amd64 and arm64. As for the copying, in the current version a value is being copied into a vector then into the iobuf. We could vectorize the copy into the iobuf, however, I don't see a way to vectorize the copy into the vector. The design purposed in this PR just has the copy into the iobuf. We could always add buffering to group values together before inserting into the iobuf; I'm doing something like it for the `boolean_value` encoder. Though I'm currently unsure if this would be better for performance.

Reply:
Quick question, the current encoding is rather basic and doesn't require much processing to convert from the in-memory value to the encoded value. Are there other encodings that we're considering that would require more processing?

Reply:
As Tyler alludes to, I think this is very far from being vectorized given the `append()`. As a rule, compiler still can't vectorize even very simple versions of something like this, like `push_back` a constant value into a vector. To vectorize this stuff, you need to use the "anti-pattern" of resizing (not reserving) the vector to the full size, then work on the default constructed elements.

Reply:
Here's an example of successful vectorization using that trick: 

https://godbolt.org/z/7s78jdvjc

The basic block at `LBB0_10` is the primary vectorized loop, copying 128 bytes per iteration using 4 SIMD load-store pairs.

Reply:
It also works for `cpu_to_be` which actually needs to do a shuffle: it is able to vectorize the shuffle.

This is all cool, but I think it probably makes a very small difference in runtime? The insertion of elements into this array (and per-element code before that) won't be vectorized so I feel like this doesn't matter much.

Reply:
Here's where I have arrived here:

* Plain encoding is so trivial it's not useful to vectorize (because we can't vectorize the input to the column writer, the record shredding).
* More complicated encodings that might be vectorizable can still be collected locally in this encoder class, then periodically "flushed" in a vectorized way to the `iobuf` layer.
* It's unlikely we're going to do more complex encodings anytime soon, so not forcing the buffering seems like just a straight win.

So I like the extra class for the encoding and going straight there because we can make local decisions about if buffering is useful or not.

Reply:
Sounds right to me.